server/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedShardsAllocatorTests.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.action.support.replication.ClusterStateCreationUtils;
+import org.elasticsearch.cluster.ClusterInfo;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ESAllocationTestCase;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.RoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.elasticsearch.cluster.routing.allocation.allocator.BalancedShardsAllocator;
+import org.elasticsearch.cluster.routing.allocation.decider.AllocationDeciders;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.snapshots.SnapshotShardSizeInfo;
+
+import java.util.Collections;
+
+public class BalancedShardsAllocatorTests extends ESAllocationTestCase {
+    public void testDecideShardAllocation() {
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(Settings.EMPTY);
+        ClusterState clusterState = ClusterStateCreationUtils.state("idx", false, ShardRoutingState.STARTED);
+        assertEquals(clusterState.nodes().getSize(), 3);
+        clusterState = allocateNew(clusterState);
+
+        ShardRouting shard = clusterState.routingTable().index("idx_new").shard(0).primaryShard();
+        RoutingAllocation allocation = new RoutingAllocation(
+            new AllocationDeciders(Collections.emptyList()),
+            new RoutingNodes(clusterState, false),
+            clusterState,
+            ClusterInfo.EMPTY,
+            SnapshotShardSizeInfo.EMPTY,
+            System.nanoTime()
+        );
+        allocation.debugDecision(true);
+
+        BalancedShardsAllocator.WeightFunction weightFunction = new BalancedShardsAllocator.WeightFunction(0.45f, 0.55f);
+        BalancedShardsAllocator.Balancer balancer = new BalancedShardsAllocator.Balancer(logger, allocation, weightFunction, 1.0f);
```

Comment:
These things are really internal details of the `BalancedShardsAllocator`. Is there a way to see the effects of this change from the outside? I.e. can you just call `BalancedShardsAllocator#allocate()` and assert something about the resulting allocation?

Reply:
I check the allocationDecision with debug on and off, and the allocationResult in the test. They should pick the same min weight node.

Reply:
Noted, but still we're not using a whole `BalancedShardsAllocator`, just these internal components. Can we test the same thing without using its internals? There are some changes to the allocator in the works and tests like this will need to be rewritten if the implementation details change.

Reply:
I still don't understand the whole `BalancedShardsAllocator` you are talking about. Can you point out that the part is the internals you are talking about?

And both `allocate()` and `decideShardAllocation()` are declared in interface `ShardsAllocator`. What kind of changes may cause the two interfaces to return different decisions? Would you mind giving me some examples? thx

Reply:
Ah sorry I think I've been seeing stale data on Github today (hence my duplicated comment earlier too). Now that I look again I see you're no longer creating a `BalancedShardsAllocator.WeightFunction` or `BalancedShardsAllocator.Balancer` directly. This looks more like what I meant. I'll do a detailed look shortly.